### PR TITLE
Feature/gsye 1 Enable exporting savings kpi for community areas 

### PR DESCRIPTION
**Pull request by hannesdiedrich**
This PR includes:
1. Removing the limit of exporting of savings KPI only for homes
2. Enable calculating of savings_percentage also for negative base costs
3. Fix treatment of battery in the savings KPI calculation 
4. Fix calculation of gsy_e_cost for producers
5. Accumulate all savings from children areas to the parent savings (which is the main task of this task)
6. Refactoring for better readability of the kpi.py module

### CODE REVIEW

**Review by Hassan754 — APPROVED** [review]

**Review by spyrostz — APPROVED** [review]

File: gsy_framework/sim_results/kpi.py [hunk 1]
```diff
@@ -151,49 +153,41 @@ def update_area_kpi(self, area_dict: Dict, core_stats: Dict):
...
-            if (trade["seller_origin_id"] in self.producer_ess_set and
-                    trade["buyer_origin_id"] not in self.consumer_ess_set):
-                self.fit_revenue += fir_excl_gf_alp * trade["energy"]
-                self.gsy_e_cost -= trade["price"]
-            if trade["buyer_origin_id"] in self.consumer_ess_set and \
-                    trade["seller_origin_id"] not in self.producer_ess_set:
-                self.utility_bill += mmr_incl_gf_alp * trade["energy"]
+            if trade["seller_origin_id"] in self.producer_ess_set:
+                if feed_in_tariff > 0:
+                    self.fit_revenue += feed_in_tariff * trade["energy"]
```
Comment (hannesdiedrich) [inline gsy_framework/sim_results/kpi.py:184]: The fit_revenue should only be calculated if the feed_in_tariff is larger than the grid_fee a long the path.

File: gsy_framework/sim_results/kpi.py [hunk 2]
```diff
@@ -151,49 +153,41 @@ def update_area_kpi(self, area_dict: Dict, core_stats: Dict):
...
+                self.gsy_e_cost -= trade["price"] - trade["fee_price"]
```
[the remark this answers is not in the document]
Comment (spyrostz) [reply gsy_framework/sim_results/kpi.py:185]: To summarise for my understanding, gsy_e_cost refers to the accumulated price balance of all the assets of the subarea. The trade["price"] though includes grid fees for each market along the path from the producer to the area of the SavingsKPI calculation, therefore in order to calculate the correct price (which is going to be the revenue of the storage essentially) we need to subtract the grid fees from it (which sounds reasonable to me BTW). Did I get it correctly?

[the remark this answers is not in the document]
Comment (hannesdiedrich) [reply gsy_framework/sim_results/kpi.py:185]: You nailed it! :-)

**Review by spyrostz — APPROVED** [review]
Some clarification comments, LGTM though since they will probably not need changes. Thanks!

File: gsy_framework/sim_results/kpi.py [hunk 3]
```diff
@@ -263,36 +255,34 @@ def memory_allocation_size_kb(self):
...
         return f"KPI: {self.performance_indices}"
 
-    def area_performance_indices(self, area_dict: Dict, core_stats: Dict) -> Dict:
-        """Entrypoint to be triggered after every market cycle to calculate
-        respective area's KPIs"""
+    def _init_states(self, area_dict: Dict):
         if area_dict["uuid"] not in self.state:
             self.state[area_dict["uuid"]] = KPIState()
+        if area_dict["uuid"] not in self.savings_state and area_dict.get("parent_uuid"):
+            # only savings KPIs for non-root areas are of interested
```
[the remark this answers is not in the document]
Comment (hannesdiedrich) [reply gsy_framework/sim_results/kpi.py:262]: Actually, this is to omit calculating the savings for the "Grid" area. KPIs are only calculated for market areas.

File: gsy_framework/sim_results/kpi.py [hunk 4]
```diff
@@ -306,60 +309,56 @@ def area_performance_indices(self, area_dict: Dict, core_stats: Dict) -> Dict:
...
-                "gsy_e_cost": area_kpis.get("gsy_e_cost"),
-                "saving_absolute": area_kpis.get("saving_absolute"),
-                "saving_percentage": area_kpis.get("saving_percentage")
-                }
-
-    # pylint: disable=arguments-renamed
+            area_kpis["self_consumption"] = 0.0
+
+        # to be consistent with the old version, where no name is present in the kpi structure
+        area_kpis.pop("name", None)
```
Comment (fievelk) [inline gsy_framework/sim_results/kpi.py:326]: Couldn't this be done by the frontend? We could send them everything and then the FE decides what to show/hide.

Comment (fievelk) [reply gsy_framework/sim_results/kpi.py:326]: Probably the entire logic of `_get_ui_formatted_results()` could/should be performed by the frontend, if it's only used to format data for the UI.

**Review by fievelk — APPROVED** [review]
One minor comment, LGTM

File: gsy_framework/sim_results/kpi.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fievelk) [reply gsy_framework/sim_results/kpi.py:262]: It might be worth adding a comment to explain :)

File: gsy_framework/sim_results/kpi.py (hunk 4 above)
Comment (fievelk) [reply gsy_framework/sim_results/kpi.py:327]: I would be up for it, yes 👍

**Review by fievelk — APPROVED** [review]

**Review by spyrostz — APPROVED** [review]
LGTM

**Pull request merged**